xla/pjrt/pjrt_stream_executor_client.cc

```diff
@@ -3448,7 +3460,32 @@ PjRtStreamExecutorClient::Compile(mlir::ModuleOp module,
       module, xla_computation,
       /*use_tuple_args=*/options.parameter_is_tupled_arguments,
       /*return_tuple=*/false));
-  return Compile(xla_computation, options);
+
+  TF_ASSIGN_OR_RETURN(std::vector<LayoutMode> arg_layout_modes,
+                      GetArgLayoutModes(module));
+  TF_ASSIGN_OR_RETURN(std::vector<LayoutMode> out_layout_modes,
+                      GetOutputLayoutModes(module));
+  TF_ASSIGN_OR_RETURN(std::vector<MemorySpaceColor> arg_memory_spaces,
+                      GetArgMemoryKinds(module));
+  TF_ASSIGN_OR_RETURN(std::vector<MemorySpaceColor> out_memory_spaces,
+                      GetOutputMemoryKinds(module));
+
+  Shape result_layout;
+  ExecutableBuildOptions build_options = options.executable_build_options;
+  TF_ASSIGN_OR_RETURN(auto arg_layouts_and_pointers,
+                      LayoutModesToXla(
+                          xla_computation, arg_layout_modes, out_layout_modes,
+                          arg_memory_spaces, out_memory_spaces,
+                          [this](Shape shape) -> absl::StatusOr<Shape> {
+                            return this->client()
+                                ->backend()
+                                .transfer_manager()
+                                ->ChooseCompactLayoutForShape(shape);
+                          },
+                          build_options));
+
+  return Compile(xla_computation, arg_layouts_and_pointers.second,
+                 *(build_options.result_layout()), options);
```

Comment:
What's `build_options.result_layout`? The out_layouts we care about are the layouts and memory_spaces you calculated on line 3470 and 3466.

Reply:
For some reason, the LayoutModesToXla function stores the result into build_options, see here: https://github.com/openxla/xla/blob/5e5a794c256b91695f90f7502185aa56cf80b66c/xla/pjrt/utils.cc#L705

The layouts calculated on line 3470 and 3466 need to be converted to XLA, that's what LayoutModesToXla function is doing.

Reply:
TBH, the LayoutModesToXla function has some weird contract, maybe we should be using  LayoutModesToXlaShapes?

Reply:
[Resolving, let's stick to LayoutModesToXla.]